File: DesignPatterns/01_GOF23/DP_examples/Lua/01_Creational/Factory/ConcreteCreator.lua

local creators = require("Creator")
-- 形状工厂
---@class ShapeFactory
ShapeFactory = {}

---@type table<ShapeKind, ICreator>
local factories = {}
-- 根据类型获取创建者
---@param kind ShapeKind
---@return ICreator
local function getFactory(kind)
    if not factories[kind] then
        if kind == ShapeKind.CIRCLE then
           factories[kind] = CircleCreator:new()
        elseif kind == ShapeKind.SQUARE then
           factories[kind] = SquareCreator:new()
        elseif kind == ShapeKind.RECTANGLE then
           factories[kind] = RectangleCreator:new()
        elseif kind == ShapeKind.BIG_CIRCLE then
           factories[kind] = BigCircleCreator:new()
        else
           factories[kind] = UnknownShapeCreator:new()
        end
    end
    return factories[kind]
end

---@param kind ShapeKind
---@return IShape
function ShapeFactory:GetShape(kind)
    local creator = getFactory(kind)
    return creator:getShape()
end

File: DesignPatterns/01_GOF23/DP_examples/CSharp/01_Creational/Factory/ConcreteCreator.cs
using System.Diagnostics.Contracts;

namespace Creational.Factory;

public partial class ShapeFactory {
    static Dictionary<ShapeKind, ICreator> Factories = new Dictionary<ShapeKind, ICreator>();
    public static IShape GetShape(ShapeKind kind) => getFactory(kind).getShape();
    static ICreator getFactory(ShapeKind kind) {
        ICreator creator = null;
        if (Factories.ContainsKey(kind))
            creator = Factories[kind];
        else {
            creator = Creator.GetCreator(kind);
            Factories.Add(kind, creator);
        }
        return creator;
    }
}
